Correct codefix by removing private modifier

KnisterPeter opened the pull request:
In case of private attribute and private constructor parameter with
assignment in the constructor body, the parameter is flagged as unused.
This is caused by the private modifier which is shadowed by the
explicity assignment in the body.
This commit updates the codefix to just remove the private modifier in
this cases.

Status:
* [x] There is an associated issue that is labeled
  'Bug' or 'help wanted' or is in the Community milestone
* [x] Code is up-to-date with the `master` branch
* [x] You've successfully run `jake runtests` locally
* [x] You've signed the CLA
* [x] There are new or updated unit tests validating the change

Fixes #24931

sheetalkamat commented on src/services/codefixes/fixUnusedIdentifier.ts:
```diff
@@ -200,8 +200,14 @@ namespace ts.codefix {
-            deleteUnusedArguments(changes, sourceFile, p, sourceFiles, checker);
+            const privateModifier = findModifier(p, SyntaxKind.PrivateKeyword);
+            if (privateModifier) {
+                changes.deleteModifier(sourceFile, p.modifiers![0]);
```
This needs to check if property is unused before this action can take place.

KnisterPeter replied:
@sheetalkamat  Can you give me a hint on how to do so? Do I have to walk the AST or are there already tools ready for that?

sheetalkamat replied:
@andy-ms to recommend better fix.

andy-ms replied:
I think we only provide an unused diagnostic if the property is unused. We won't provide a diagnostic when the parameter itself is unused because it is at least used to set the property.
I think the condition should be if `p.modifiers` exists at all, and it should delete all modifiers rather than just handling `private`; any other modifiers (`public`, `protected`, `readonly`) also make this a property.

KnisterPeter replied:
@andy-ms As mentioned in the related issue I'm adding the other modifiers.
The main question is if its okay to remove just the modifiers. There are cases where the property is flagged as unused, and if the modifiers are removed it may still be unused since the parameter is also unused.
Would a two-step fixing be okay? Or should it be in just one step?

KnisterPeter commented:
@sheetalkamat @andy-ms I've updated the PR and added `public`, `protected` and `readonly` next to `private`. All modifiers are removed as soon as the quickfix is executed.

KnisterPeter commented on tests/cases/fourslash/codeFixUnusedIdentifier_parameter_modifier.ts:
```diff
@@ -0,0 +1,22 @@
+////        this.prop = arg;
+////    }
+////}
+
```
@sheetalkamat Right now the warning does not differ if the property is unused or if the parameter is unused. Therefore I asked for a unitlity function which can give me this information and descide upon that. Can you hint me somewhere?

andy-ms commented on src/services/codefixes/fixUnusedIdentifier.ts:
```diff
@@ -200,8 +200,17 @@ namespace ts.codefix {
         if (mayDeleteParameter(p, checker, isFixAll)) {
-            changes.delete(sourceFile, p);
-            deleteUnusedArguments(changes, sourceFile, p, sourceFiles, checker);
+            const modifiers: Modifier["kind"][] = [SyntaxKind.PrivateKeyword, SyntaxKind.ProtectedKeyword, SyntaxKind.PublicKeyword, SyntaxKind.ReadonlyKeyword];
```
Think this would be simpler to just check for the presence of *any* modifiers and delete them all at once. Since these are the only legal modifiers anyway.

andy-ms commented on src/services/codefixes/fixUnusedIdentifier.ts:
```diff
@@ -200,8 +200,17 @@ namespace ts.codefix {
-            deleteUnusedArguments(changes, sourceFile, p, sourceFiles, checker);
+            const modifiers: Modifier["kind"][] = [SyntaxKind.PrivateKeyword, SyntaxKind.ProtectedKeyword, SyntaxKind.PublicKeyword, SyntaxKind.ReadonlyKeyword];
+            const foundModifiers = modifiers.map(modifier => findModifier(p, modifier)).filter(modifier => modifier !== undefined);
+            if (foundModifiers.length > 0) {
```
This could test `foundModifiers.length > 0 && (!isIdentifier(p.name) || FindAllReferences.Core.isSymbolReferencedInFile(p.name, checker, sourceFile))` so for a completely unused parameter you delete it as usual.

KnisterPeter commented:
@sheetalkamat @andy-ms Update for your comments. Can you please review again?

andy-ms commented:
You can fix that by changing the first line of `eachSymbolReferenceInFile` to:
```ts
const symbol = isParameterPropertyDeclaration(definition.parent)
    ? first(checker.getSymbolsOfParameterPropertyDeclaration(definition.parent, definition.text))
    : checker.getSymbolAtLocation(definition);
```
You'll also have to update `deleteAssignments` to have more checks:
```ts
FindAllReferences.Core.eachSymbolReferenceInFile(token, checker, sourceFile, (ref: Node) => {
    if (isPropertyAccessExpression(ref.parent) && ref.parent.name === ref) ref = ref.parent;
    if (isBinaryExpression(ref.parent) && isExpressionStatement(ref.parent.parent) && ref.parent.left === ref) {
        changes.delete(sourceFile, ref.parent.parent);
    }
});
```

KnisterPeter commented:
@andy-ms Awesome, that did work. Thank you. I've rebased to the latest master now.
@sheetalkamat  Please rereview.
